README.md
#peddlr

peddlr helps you plan your Citi Bike trips around New York City.

#Description

Citi Bike is a system of publically available bikes placed at stations throughout New York City. But planning a trip using Citi Bike can be complicated. Where is the closest station? Is it active? Does it have any bikes availabe?

peddlr helps you answer these questions and allows you to plan you trips around the city. Users plan a trip by entering origin and destination addresses then peddlr  looks up the nearest Citi Bike station with available bikes and provides usres with a map to the station. peddlr also directs you to the Citi Bike station nearest your destination. In addition, peddlr allows you to look at past trips you've gone on.

Take the confusion out of Citi Bike. User peddlr.

#Screenshots

![screenshot](https://dl.dropboxusercontent.com/u/1361162/Screen%20Shot%202014-05-22%20at%209.12.07%20AMEDT.jpg)

![screenshot](https://dl.dropboxusercontent.com/u/1361162/Screen%20Shot%202014-05-22%20at%204.21.09%20PMEDT.jpg)

#Code Sample

```ruby

def self.nearest_station(address_coordinates)
  all_stations = Citibikenyc.branches["results"]
  station_array = []

  all_stations.each do |station|

    distance_from_address = Geocoder::Calculations.distance_between(address_coordinates, [(station['latitude']),(station['longitude'])])

    station_info = { 
      :station_id => (station['id']), 
      :label => (station['label']), 
      :latitude => (station['latitude']),
      :longitude => (station['longitude']),
      :distance => distance_from_address 
    }

    station_array << station_info
  end

  sorted = station_array.sort{ |x, y| x[:distance] <=> y[:distance] }

  sorted.each do |station|
    return station if station_state(station[:station_id]).fetch("status", "not active") == "Active" &&  station_state(station[:station_id]).fetch("availableBikes", 0).to_i > 0
  end

end


```

###RESOURCES

#### Trello

- https://trello.com/b/Q5YkldhD/bike-app

#### Heroku

- http://thawing-atoll-1044.herokuapp.com/

#### Citi Bike System Data

- http://www.citibikenyc.com/system-data

#### Citibikenyc gem

- gets realtime data on Citi Bike stations throughout the city

- https://github.com/edgar/citibikenyc

- Get a list of all stations
-- Citibikenyc.stations

- Get only the current status for all stations
-- Citibikenyc.stations_status

- Get helmets info
-- Citibikenyc.helmets

- Get branches info
-- Citibikenyc.branches

#### Geocoder gem

- http://www.rubygeocoder.com/
- http://rdoc.info/github/alexreisner/geocoder/master/frames

- uses Google Maps API to allow for various forms of spatial analysis, including distance calculations between two location and geocoding street address to lat/long

##### Useful commands for Geocoder

- look up coordinates of some location (like searching Google Maps)

```ruby
Geocoder.coordinates("25 Main St, Cooperstown, NY")
 => [42.700149, -74.922767]
 ```

- distance (in miles) between Eiffel Tower and Empire State Building

```ruby
Geocoder::Calculations.distance_between([47.858205,2.294359], [40.748433,-73.985655])
 => 3619.77359999382
```





CHANGELOG.md
##CHANGE LOG

#### 5/15/14

* created rails app for bikes
* got gems for geocoder and citibikesnyc
* set up routes, controllers, models and views for CRUD app
* set up migrations 
* seeded STATIONS table from Citibikenyc.branches method, now populated with data on all Citi Bike stations throughout the city

####5/16/14

* changed database structure; added separate table for locations to jib with Geocoder gem
* changed db structure back to original, dropped locations table, figured out how to use geocoder methods etc.
* worked on logic for calculating distance between entered address and nearest Citi Bike stations
* added method "nearest" in trip model

####5/17/14

* add radio button to address form for current location
* added columns to trips table
* reformatted new form for trips

####5/19/14

* added user and authorization functionality
* got trip form working so that new trips can be created
* worked on trips#show and trips#index pages
* got user authorization working

####5/20/14

* got app working on heroku - thawing-atoll-1044.herokuapp.com
* got IP address location working
* nearest station results now filter out stations that are inactive or that have 0 bikes
* address address label using geocoder by reverse geocoding the coordinates
* got validation working for sign-up

####5/21/14

* got validation working on update
* added google map static image with drop points for stations and addresses and simple travel path
* removed stations table from database







app/models/trip.rb
class Trip < ActiveRecord::Base

  belongs_to :user



  def self.new_trip(start_address, end_address)
    start_coordinates = address_coordinates(start_address)
    end_coordinates = address_coordinates(end_address)

    start_station = nearest_station(start_coordinates)
    end_station = nearest_station(end_coordinates)

    start_address_label = coordinates_address(start_coordinates)
    end_address_label = coordinates_address(end_coordinates)

    trip = trip_hash(start_station, end_station, start_address, end_address, start_coordinates, end_coordinates, start_address_label, end_address_label)

    self.create(trip)
    
  end



  def self.address_coordinates(address)
    Geocoder.coordinates(address)
  end

  def self.coordinates_address(coordinates)
    Geocoder.address(coordinates)
  end



  def self.nearest_station(address_coordinates)
    all_stations = Citibikenyc.branches["results"]
    station_array = []

    all_stations.each do |station|

      distance_from_address = Geocoder::Calculations.distance_between(address_coordinates, [(station['latitude']),(station['longitude'])])

      station_info = { 
        :station_id => (station['id']), 
        :label => (station['label']), 
        :latitude => (station['latitude']),
        :longitude => (station['longitude']),
        :distance => distance_from_address 
      }

      station_array << station_info
    end

    sorted = station_array.sort{ |x, y| x[:distance] <=> y[:distance] }

    sorted.each do |station|
      return station if station_state(station[:station_id]).fetch("status", "not active") == "Active" &&  station_state(station[:station_id]).fetch("availableBikes", 0).to_i > 0
    end

  end

 


  def self.trip_hash(start_station, end_station, start_address, end_address, start_coordinates, end_coordinates, start_address_label, end_address_label)
    {
      :start_station_id => start_station[:station_id],
      :start_station_label => start_station[:label],
      :start_station_latitude => start_station[:latitude],
      :start_station_longitude => start_station[:longitude],
      :end_station_id => end_station[:station_id],
      :end_station_label => end_station[:label],
      :end_station_latitude => end_station[:latitude],
      :end_station_longitude => end_station[:longitude],
      :start_address => start_address,
      :start_lat => start_coordinates[0],
      :start_long => start_coordinates[1],
      :end_address => end_address,
      :end_lat => end_coordinates[0],
      :end_long => end_coordinates[1],
      :start_address_label => start_address_label,
      :end_address_label => end_address_label
    }
  end



  def total_distance
    ( Geocoder::Calculations.distance_between([self.start_station_latitude, self.start_station_longitude], [self.start_lat, self.start_long]) + 
      Geocoder::Calculations.distance_between([self.start_lat, self.start_long], [self.end_lat, self.end_long]) + 
      Geocoder::Calculations.distance_between([self.end_lat, self.end_long], [self.end_station_latitude, self.end_station_longitude]) 
    ).round(2)
  end


  private 

  def self.station_state(id)
    station_activity_arr = Citibikenyc.stations_status["results"]
    station_activity_arr.each do |station|
      return station if station.id == id
    end
    {status: "Not Active"}
  end

end




# Citibikenyc.stations_status["results"]
# ___________________________
# [331] {
#                   "id" => 3002,
#               "status" => "Active",
#       "availableBikes" => 17,
#       "availableDocks" => 6
#   }


# bike_stations
# _____________
# {
#   :start => {
#     :station_id => 127,
#     :label => "One Court Square",
#     :latitude => 40.7467393577099,
#     :longitude => -73.9434874802828,
#     :distance => 0.5966593465631346
#     },
#     :end => {
#       :station_id => 228,
#       :label => "22 W. 32nd St",
#       :latitude => 40.7475128173828,
#       :longitude => -73.986946105957,
#       :distance => 0.09803588893167761
#     }
#   }







